src/librustdoc/doctest.rs

```diff
@@ -446,18 +527,41 @@ fn run_test(
         }
     }
 
-    compiler.arg("-");
-    compiler.stdin(Stdio::piped());
-    compiler.stderr(Stdio::piped());
+    // If this is a merged doctest, we need to write it into a file instead of using stdin
+    // because if the size of the merged doctests is too big, it'll simply break stdin.
+    if is_multiple_tests {
+        // It makes the compilation failure much faster if it is for a combined doctest.
+        compiler.arg("--error-format=short");
+        let input_file =
+            doctest.test_opts.outdir.path().join(&format!("doctest_{}.rs", doctest.edition));
```

Comment:
It would be good to have a few utility functions that'd generate the names of the merged doctests. That way we avoid magic constants that could fall out of date with each other.

Something like
```rust
fn filename_for_merged_doctest(edition: Edition) -> String {format!(...)}
```

Reply:
I made it into a method that returns the path.